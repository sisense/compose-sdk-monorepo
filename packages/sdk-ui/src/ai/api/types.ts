import { JaqlQueryPayload } from '@sisense/sdk-query-client';

import { ExpandedQueryModel } from '@/analytics-composer';
import { WidgetProps } from '@/props';

export interface ChatContext {
  title: string;
  live: boolean;
}

export interface ChatContextDetails {
  dashboardId?: string;
}
export interface TextMessage {
  role: 'assistant';
  content: string; // stringified json { answer: string }
  type: 'text';
}
export interface NlqMessage {
  role: 'assistant';
  content: string; // stringified json
  type: 'nlq';
}
interface RegularMessage {
  role: 'user' | 'assistant';
  content: string;
}
export type ChatMessage = RegularMessage | NlqMessage | TextMessage;

export interface Chat {
  chatId: string;
  contextId: string;
  contextTitle: string;
  contextDetails?: ChatContextDetails;
  expireAt: string;
  tenantId: string;
  userId: string;
  chatHistory: ChatMessage[];
}
export type ChatWithoutHistory = Omit<Chat, 'chatHistory'>;

export interface LlmConfig {
  provider: string;
  apiKey: string;
  baseUrl: string;
  model: string;
  version?: string;
}

export interface ChatRequest {
  text?: string;
  timezone?: string;
  options?: {
    enableFollowup?: boolean;
  };
}

interface TextResponse {
  chatHistory: ChatMessage[];
  data: {
    answer: string;
  };
  chatId: string;
  responseType: 'text';
}

interface ErrorResponse {
  data: {
    message: string;
    code: string;
    answer: string;
  };
  responseType: 'error';
}

/**
 * Response data for NLQ request
 */
export interface NlqResponseData extends ExpandedQueryModel {
  /** Detailed description of the response */
  detailedDescription: string;
  /** An array of suggested followup questions  */
  followupQuestions?: string[];
  /** NLQ prompt used in the request  */
  nlqPrompt: string;
  /** Suggested widget props*/
  widgetProps?: WidgetProps | undefined;
  /** The response message for the chat  */
  userMsg: string;
  /** @internal */
  clarification?: string;
}
export interface NlqResponse {
  data: NlqResponseData;
  timestamp: string;
  responseType: 'nlq';
}
export type ChatResponse = NlqResponse | TextResponse | ErrorResponse;

export interface QueryRecommendationConfig {
  numOfRecommendations: number;
  userPrompt?: string;
}

export type NlqResult = Omit<NlqResponseData, 'followupQuestions'>;

/** AI-generated recommended query you can run on your data model */
export interface QueryRecommendation extends Omit<NlqResponseData, 'followupQuestions'> {}
export type QueryRecommendationResponse = QueryRecommendation[];

export interface GetNlgInsightsRequest {
  jaql: Pick<JaqlQueryPayload, 'datasource' | 'metadata' | 'filterRelations'>;
  verbosity?: 'Low' | 'High';
}

/** @internal */
export interface GetNlgInsightsResponse {
  data?: {
    answer: string;
  };
  responseType: 'Text';
}

/** @internal */
export const NLQ_RESULT_CHART_TYPES = [
  'indicator',
  'column',
  'line',
  'table',
  'pie',
  'bar',
] as const;
/** NLQ result chart types @expandType */
export type NlqResultChartType = (typeof NLQ_RESULT_CHART_TYPES)[number];
export interface GetNlqResultRequest {
  text: string;
  timezone?: string;
  chartTypes: NlqResultChartType[];
}

export interface SendFeedbackRequest {
  type: string;
  data: object;
  sourceId: string;
  rating: -1 | 1;
}

/**
 * The chat mode to use for a chat session
 *
 * analyze: Enable business users to uncover data insights
 * develop: Enable developers to build queries and charts for embedded analytics
 */
export type ChatMode = 'analyze' | 'develop';
