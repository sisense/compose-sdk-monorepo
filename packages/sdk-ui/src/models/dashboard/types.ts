import { CommonFiltersOptions } from '@/common-filters/types';
import { ColorPaletteTheme, TabberConfig } from '@/types';

export type {
  CommonFiltersOptions,
  CommonFiltersApplyMode,
  FiltersIgnoringRules,
} from '@/common-filters/types';

/**
 * Column layout of dashboard widgets
 */
export interface WidgetsPanelColumnLayout {
  columns: {
    widthPercentage: number;
    rows: {
      cells: {
        widthPercentage: number;
        /**
         * @privateRemarks
         * This value appears to be overwritten by the widget's height property.
         * Marking as optional and internal for now.
         * @internal
         */
        height?: number | string;
        widgetId: string;
      }[];
    }[];
  }[];
}

/**
 * Layout of dashboard widgets panel, which is a union of different layout algorithms
 */
export type WidgetsPanelLayout = WidgetsPanelColumnLayout;

/**
 * Widget ID
 */
export type WidgetId = string;
/**
 * Options for widgets in a dashboard
 *
 * For example, how common filters defined at the dashboard level should be applied to widgets.
 */
export type WidgetsOptions = Record<WidgetId, { filtersOptions?: CommonFiltersOptions }>;
/**
 * Options for TabberWidetss in a dashboard
 *
 * This property actually moves responsibility on the layout management from the tabber widgets to the dashboard,
 * storing all the tabbers configs in the single place
 * @internal
 */
export type TabbersOptions = Record<WidgetId, TabberConfig>;

/**
 * Style options for the dashboard.
 */
export type DashboardStyleOptions = {
  /** Collection of colors used to color various elements */
  palette?: ColorPaletteTheme;
  /** Background color */
  backgroundColor?: string;
  /** Width of the divider line between widgets */
  dividerLineWidth?: number;
  /** Divider line color */
  dividerLineColor?: string;
};
