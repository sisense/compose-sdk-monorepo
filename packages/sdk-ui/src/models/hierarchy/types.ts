import type { Attribute, DataSource } from '@sisense/sdk-data';

/**
 * Options for retrieving hierarchies.
 *
 * @internal
 */
export type GetHierarchiesOptions = {
  /**
   * The dimension for which to retrieve the hierarchies.
   */
  dimension: Attribute;
  /**
   * The data source from which to retrieve the hierarchies - e.g. `Sample ECommerce`.
   *
   * If not specified, the query will use the `defaultDataSource` specified in the parent Sisense Context.
   */
  dataSource?: DataSource;
  /**
   * A list of hierarchy IDs to retrieve specific hierarchies.
   */
  ids?: string[];
  /**
   * A flag indicating whether to filter the retrieved hierarchies based on the `alwaysIncluded` field.
   *
   * When set to true, only hierarchies with `alwaysIncluded: true` will be returned.
   * When set to false, only hierarchies with `alwaysIncluded: false` will be returned.
   * If not specified, all hierarchies will be returned.
   */
  alwaysIncluded?: boolean;
};
