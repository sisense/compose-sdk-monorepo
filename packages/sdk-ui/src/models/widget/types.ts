import { ChartType } from '@/index';

/**
 * Widget type
 *
 * todo: as a future refactoring need to replace this type by combination of WidgetType and ChartType
 *
 * @internal
 */
export type WidgetTypeInternal = ChartType | 'pivot' | 'plugin' | 'text';
