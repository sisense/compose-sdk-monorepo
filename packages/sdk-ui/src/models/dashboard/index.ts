export * from './types';
export * from './get-dashboard-model';
export * from './get-dashboard-models';
export * from './use-get-dashboard-model';
export * from './use-get-dashboard-models';
export * from './dashboard-model';
export * as dashboardModelTranslator from './dashboard-model-translator';
