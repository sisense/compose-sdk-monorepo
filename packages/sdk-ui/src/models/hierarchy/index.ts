export * from './hierarchy-model';
export * as hierarchyModelTranslator from './hierarchy-model-translator';
export * from './use-get-hierarchy-models';
export * from './types';
export * from './get-hierarchy-models';
