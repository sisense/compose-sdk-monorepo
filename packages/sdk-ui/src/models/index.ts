export * from './dashboard';
export * from './widget';
