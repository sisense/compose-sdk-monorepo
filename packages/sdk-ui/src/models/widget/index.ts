export * from './types';
export * from './translate-widget';
