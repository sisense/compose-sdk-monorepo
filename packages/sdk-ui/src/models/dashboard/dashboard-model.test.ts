import { dashboardModelTranslator } from '@/models';
import { WidgetStyle } from '@/widget-by-id/types';
import { DataType } from '@sisense/sdk-data';

describe('DashboardModel', () => {
  it('should create a new dashboard model with minimum of arguments', async () => {
    expect(
      dashboardModelTranslator.fromDashboardDto({
        oid: 'test',
        title: 'test',
        datasource: { title: 'test', id: 'test' },
      }),
    ).toMatchSnapshot();
  });

  it('should create a new dashboard model with all arguments', () => {
    expect(
      dashboardModelTranslator.fromDashboardDto({
        oid: 'test',
        title: 'test',
        datasource: { title: 'test', id: 'test' },
        widgets: [
          {
            oid: 'widget-1',
            type: 'chart/pie',
            subtype: 'pie/donut',
            datasource: {
              title: 'test',
              id: 'test',
            },
            metadata: {
              panels: [],
            },
            style: {} as WidgetStyle,
            title: 'test',
            desc: null,
          },
        ],
        layout: {
          columns: [
            {
              width: 1000,
              cells: [
                {
                  subcells: [
                    {
                      width: 2000,
                      elements: [
                        {
                          height: 3000,
                          widgetid: 'widget-1',
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
        },
        filters: [
          {
            jaql: {
              datatype: DataType.TEXT,
              dim: '[table.column]',
              table: 'table',
              column: 'column',
              title: 'title',
              filter: {
                members: ['member'],
              },
            },
            instanceid: 'test-id',
            disabled: false,
          },
        ],
        settings: {
          autoUpdateOnFiltersChange: true,
          useAcceleration: true,
          aiAssistantEnabled: true,
          managedByTool: 'studio-assistant',
        },
      }),
    ).toMatchSnapshot();
  });
});
