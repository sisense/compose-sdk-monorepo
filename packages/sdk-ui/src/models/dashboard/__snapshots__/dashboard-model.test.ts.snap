// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DashboardModel > should create a new dashboard model with all arguments 1`] = `
DashboardModel {
  "dataSource": {
    "title": "test",
    "type": "elasticube",
  },
  "filters": [
    {
      "__serializable": "DimensionalElement",
      "attribute": {
        "__serializable": "DimensionalElement",
        "desc": "",
        "expression": "dim",
        "name": "column",
        "type": "text-attribute",
      },
      "desc": "",
      "filterType": "members",
      "members": [
        "member",
      ],
      "name": "e8e9681675b",
      "type": "filter",
    },
  ],
  "layout": {
    "columns": [
      {
        "rows": [
          {
            "cells": [
              {
                "height": 3000,
                "widgetId": "widget-1",
                "widthPercentage": 2000,
              },
            ],
          },
        ],
        "widthPercentage": 1000,
      },
    ],
  },
  "oid": "test",
  "styleOptions": {},
  "title": "test",
  "widgetFilterOptions": {
    "widget-1": {
      "applyMode": "highlight",
      "forceApplyBackgroundFilters": true,
      "ignoreFilters": {
        "all": undefined,
        "ids": undefined,
      },
      "shouldAffectFilters": undefined,
    },
  },
  "widgets": [
    WidgetModel {
      "chartType": "pie",
      "dataOptions": {
        "category": [],
        "value": [],
      },
      "dataSource": "test",
      "description": "",
      "drilldownOptions": {
        "drilldownDimensions": [],
        "drilldownSelections": [],
      },
      "filters": [],
      "oid": "widget-1",
      "styleOptions": {
        "subtype": "pie/donut",
      },
      "title": "test",
      "widgetType": "chart/pie",
    },
  ],
}
`;

exports[`DashboardModel > should create a new dashboard model with minimum of arguments 1`] = `
DashboardModel {
  "dataSource": {
    "title": "test",
    "type": "elasticube",
  },
  "filters": [],
  "layout": {
    "columns": [],
  },
  "oid": "test",
  "styleOptions": {},
  "title": "test",
  "widgetFilterOptions": {},
  "widgets": [],
}
`;
